src/Utilities/Extensions/IEnumerableExtensions.cs

```diff
@@ -71,6 +71,29 @@ public static IEnumerable<T> OrderBy<T>(this IEnumerable<T> source, Comparison<T
             return source.Where((Func<T, bool>)s_notNullTest);
         }
 
+        public static ImmutableArray<TSource> WhereAsArray<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> selector)
+        {
+            var builder = ImmutableArray.CreateBuilder<TSource>();
+            bool any = false;
+            foreach (var element in source)
+            {
+                if (selector(element))
+                {
+                    any = true;
+                    builder.Add(element);
+                }
+            }
+
+            if (any)
```

Comment:
Were you trying to make an optimization here? `ToImmutable()` doesn't allocate if the builder has no items, since empty arrays can be cached.

If you want to optimize this method, it would be better to lazy-initialize the builder to avoid the builder allocation:

```cs
ImmutableArray<>.Builder builder = null;

foreach (var element in source)
{
    if (selector(element))
    {
        builder = builder ?? IA.CreateBuilder<>();
        builder.Add(element);
    }
}

return builder?.ToImmutable() ?? ImmutableArray<TSource>.Empty;
```

Reply:
I copied this from Roslyn's implementation, with the exception that Roslyn's only works on ImmutableArray inputs.